Client- and serverside bug fixes

steforster opened the pull request:
I picked Alchemy as communication framework for [my remote actor library](https://github.com/steforster/Remact.Net) because of Alchemys slim design and the availability of a client implementation.  
But after some initial success I ran into more and more bugs, especially on the client side.
This pull request collects all committs that were needed to make the library much more stable.
When reading the list of open issues I think some of them could be fixed by this pull request too:
E.g. #93, #89, #84, #78, #72, #66, #60, #53, #49, #25  

The changed points are:
- Client connect and disconnect sequence:  
  Fixed errors and made it rfc6455 conformant
- Receive successive messages on client or server side:  
  Fixed message loss, when several messages are received in the same TCP packet
- Tested under Monodevelop
- Integration tests: New tests for successive and larger messages
- Added configuration flag "Handler.FastDirectSendingMode":  
  Useful when a server must be very responsive.  When setting this flag,
  the maximal throughput rises from 33 to more than 5000 request/response pairs per second
- Improved server side asynchrony when listening for clients and when receiving a client message.
- Removed static allocated threads on client side. These threads were dispensable in an asynchronous design.

segor committed:
added .user into ignore list

segor committed:
fixed  context early disposal  in WebSocketClient

steforster committed:
Merge pull request #1 from segor/master

merge segor's fork

steforster committed:
fixed WebSocketClient.Disconnect().
Avoids null-reference exception, when underlying socket is not connected.

steforster committed:
Socket disconnect event is sent to user code on client side.
Used to indicate when server shuts down.

steforster committed:
Made compileable under Monodevelop 2.8.6.3
Added some settings in csproj and .gitignore.

steforster committed:
Fixed opcode for close frame in WebSocketClient

steforster committed:
Added Handler.FastDirectSendingMode.

When FastDirectSendingMode is set to true, the send-threads are not started.
Messages are then sent directly from the multithreaded application to
the underlaying socket buffer. 
The Send method may block a short time
until the previous send operation has copied its data do the socket buffer.
There is a considerable speed increase for applications needing
fast roundtrip times and have a small number of clients:
    FastDirectSendingMode = false:   33 request+responses per second - caused by a delay of 10ms in every dequeue operation.
    FastDirectSendingMode = true:  2500 request+responses per second

steforster committed:
Skip pending send operations when disconnecting a client.
Removed unused method 'DoReceive'.

steforster committed:
Added static methods to shutdown Alchemy threads.
Documented the strange behavour of WebSocketServer.Dispose().

steforster committed:
Fixed client disconnect sequence.

steforster committed:
Added CancellationSource to context;
Added asynchronous connect operation to client;
Added LatestException to UserContext.
This improves WebSocketClient connect and disconnect.

steforster committed:
Corrected CR/LF.

steforster committed:
WebSocketClient: prevented null exception on disconnect.

steforster committed:
Fixed threading issues when receiving many messages in client. And when sending many messages in server or client.

steforster committed:
Fixed integration tests ClientServerSubProtocol.

steforster committed:
Support successive messages in the same receive buffer (server and client side).

steforster committed:
Support breakpoints when starting unit tests from visual studio.

steforster committed:
Fixed close message from server to client.
Support partially received headers on server and client side.
Fixed reset of FrameHeader.IsEnd.
Fixed client disconnect (make it ready for next connect).
Fixed WebSocketServer.Clients (count).
Fixed and improved integration test ClientSendDataConcurrent.
Added integration test ClientSendMultipleMessages.

steforster committed:
Improved TCPServer: Use asynchronous BeginAcceptTcpClient instead of looping in a threadpool thread.

steforster committed:
Changed serverside StartReceive(). Made it really asynchronous.
   No need to ReceiveReady.Wait until a request has been received.
Removed client side setup threads. Direct setup is faster.
   No need for static threads on client side.
Improved integration tests.

tluyben commented:
Ah this fixes that weird https://github.com/Olivine-Labs/Alchemy-Websockets/blob/master/src/Alchemy/WebSocketClient.cs#L151 stuff; I'm not sure why that actually would ever work. Thanks, please merge.

zyo2012 commented:
I've tested this fork it's running supper great, no missing packet anymore. Please merge!

steforster committed:
Reduced accessibility and made it more CLS compliant.
Added stream API to DataFrame.
Fixed UnitTests for MonoDevelop.

steforster committed:
Define EOL for Visual Studio in .gitattributes

steforster committed:
Consequences of setting crlf in .gitattributes

steforster committed:
Monodevelop line ending = native

steforster committed:
Some more file changes because of crlf.

steforster committed:
Handle too long messages when header is received: Disconnect.
FlashAccessPolicyEnabled is set through the WebSocket constructor.

steforster committed:
Made it CLS compliant. Cleaned up Context.Disconnect and Close.
Prepared stream API on DataFrame for MsgPack integration.

steforster committed:
Prevent double OnDisconnected calls.

steforster committed:
Tested DataFrame stream reading.

steforster committed:
Increased default buffersize to 1500 bytes and framesize to 1 MB.

steforster committed:
FrameBuffer supports writing as stream.

steforster committed:
Added UserContext.Disconnect(): Let the server disconnect a client.

steforster committed:
Removed dependency to System.Web.dll since it is not available in Xamarin.Android
or NET 4 Client Profile.
System.Net contains a compatible name-value-CookieCollection.

steforster committed:
Protect from NullReferenceException during shutdown or unexpected message on client.

steforster committed:
Upgraded solution to Visual Studio 2015

steforster committed:
Typo

shinigami-de committed:
Update WebSocketClient.cs

Added a ClientAddress update since it never got set due to Context initialization with "null,null"

steforster committed:
Merge pull request #2 from shinigami-de/patch-1

Update WebSocketClient.cs
